Sources/MapboxNavigation/SpriteRepository.swift

```diff
@@ -4,79 +4,93 @@ import MapboxCoreNavigation
 class SpriteRepository {
-    // Caching the Sprite and single shield icon images.
-    let spriteCache = ImageCache()
-    // Caching the single legacy shield icon images.
-    let legacyCache = ImageCache()
-    // Caching the metadata info for Sprite.
-    let infoCache =  SpriteInfoCache()
     let baseURL: URL = URL(string: "https://api.mapbox.com/styles/v1")!
-    var styleURI: StyleURI = .navigationDay
+    let defaultStyleURI: StyleURI = .navigationDay
     fileprivate(set) var imageDownloader: ReentrantImageDownloader
     
+    let requestCache: URLCaching
+    let derivedCache: BimodalImageCache
     static let shared = SpriteRepository.init()
     
     private let requestTimeOut: TimeInterval = 10
     
-    var styleID: String? {
-        styleURI.rawValue.components(separatedBy: "styles")[safe: 1]
-    }
-    
     var sessionConfiguration: URLSessionConfiguration {
         didSet {
             imageDownloader = ImageDownloader(sessionConfiguration: sessionConfiguration)
         }
     }
     
+    var userInterfaceIdiomStyles = [UIUserInterfaceIdiom: StyleURI]()
+
     init() {
         sessionConfiguration = URLSessionConfiguration.default
         sessionConfiguration.timeoutIntervalForRequest = self.requestTimeOut
         imageDownloader = ImageDownloader(sessionConfiguration: sessionConfiguration)
+        requestCache = URLDataCache()
+        derivedCache = ImageCache()
     }
     
-    func updateStyle(styleURI: StyleURI?, completion: @escaping ImageDownloadCompletionHandler) {
+    func updateStyle(styleURI: StyleURI?,
+                     idiom: UIUserInterfaceIdiom = .phone,
+                     completion: @escaping ImageDownloadCompletionHandler) {
         // If no valid StyleURI provided, use the default Day style.
-        let newStyleURI = styleURI ?? self.styleURI
-        guard newStyleURI != self.styleURI || getSpriteImage() == nil else {
+        let newStyleURI = styleURI ?? defaultStyleURI
+        
+        if !userInterfaceIdiomStyles.values.contains(newStyleURI) {
+            updateUserInterfaceIdiomStyles(styleURI: newStyleURI, idiom: idiom)
+            updateSprite(styleURI: newStyleURI, completion: completion)
+        } else {
+            updateUserInterfaceIdiomStyles(styleURI: newStyleURI, idiom: idiom)
             completion(nil)
             return
         }
-        
-        spriteCache.clearMemory()
-        infoCache.clearMemory()
-        
-        // Update the styleURI just after the Sprite memory reset. When the connection is poor, the next round of style update
-        // or the representation update could use the correct ones.
-        self.styleURI = newStyleURI
-        updateSprite(completion: completion)
+    }
+    
+    func updateUserInterfaceIdiomStyles(styleURI: StyleURI, idiom: UIUserInterfaceIdiom) {
+        if let previousStyleURI = userInterfaceIdiomStyles[idiom],
+           previousStyleURI != styleURI {
+            userInterfaceIdiomStyles[idiom] = styleURI
```

Comment:
Can we call `userInterfaceIdiomStyles[idiom] = styleURI` once at the beginning of this method?

Reply:
Yes, the previous way to call `updateUserInterfaceIdiomStyles(styleURI: idiom:)` is to remove the previous style used by the `UIUserInterfaceIdiom`, if it's not used by other `UIUserInterfaceIdiom`. We removed the unused style cache because we used to use memory. but right now we have the URLCache with capacity limits, we could keep the old style cache incase the style change and re-use it in offline mode. So `userInterfaceIdiomStyles[idiom] = styleURI` will be used to replaced the `updateUserInterfaceIdiomStyles(styleURI: idiom:)`.